### PR TITLE
Sample method determinism

### PR DESCRIPTION
**Pull request by han-ol**
This PR addresses issue #666.

Essentially, it extends the seed argument that already exists in all `keras.random` functions (like https://github.com/keras-team/keras/blob/v3.14.0/keras/src/random/random.py#L5) all the way up to the `workflow.sample` method in a way that makes sure that randomness in all different components of high-level objects (like workflow) stays uncorrelated.

Moreover: ports `distribution.Mixture` randomness from numpy to keras, adds a `keras_utils.keras_multinomial` convenience function and fixes a bug regarding correct handling of `batch_shape` between `ScoringRuleApproximator` and `MixtureScore`.

To confirm that samples for the same seed are identical, corresponding integration tests with 

* an `EnsembleApproximator` with a CouplingFlow and a DiffusionModel approximator and 
* a `ScoringRuleApproximator` with both a standalone `MvNormalScore` and a `MixtureScore` of two `MvNormalScore`s 

are passing.

to do:

* [x] The tests included in the initial commits are still "work in progress" and should be improved to run faster (especially `DiffusionModel` can be wasteful to sample) and integrate well with the existing fixtures. (The `DiffusionModel` tests were only performed locally at this point)
* [x] `workflow.sample`

### CODE REVIEW

**Comment by han-ol** [comment]
Now, includes fast tests for all inference networks. Ready for review! 

I used a kwargs.pop pattern in _inverse methods. We can also explicitly add seed to the arguments, if you prefer.

File: bayesflow/approximators/continuous_approximator.py [hunk 1]
```diff
@@ -229,6 +231,10 @@ def sample(
         return_summaries: bool, optional
             If set to True and a summary network is present, will return the learned summary statistics for
             the provided conditions.
+        seed : int, keras.random.SeedGenerator, or None, optional
+            Seed for reproducible sampling. An integer is converted to a ``keras.random.SeedGenerator``
+            and shared across all stochastic operations in the call. A ``SeedGenerator`` is passed through
+            as-is. If ``None`` (default), each component uses its own instance seed generator or a global one.
```
Comment (LarsKue) [inline bayesflow/approximators/continuous_approximator.py]: The doc is unspecific here, is a single global seed generator used or are multiple used? By default, I believe keras uses a single global one.

File: bayesflow/approximators/scoring_rule_approximator.py [hunk 2]
```diff
@@ -62,8 +63,8 @@ def __init__(
 
         self.distribution_keys = []
         for score_key, score in self.inference_network.scoring_rules.items():
-            has_sample = callable(getattr(score, "sample", None))
-            has_log_prob = callable(getattr(score, "log_prob", None))
+            has_sample = getattr(score, "sample", None)
+            has_log_prob = getattr(score, "log_prob", None)
```
Comment (LarsKue) [inline bayesflow/approximators/scoring_rule_approximator.py]: Why should these `has_*` variables not be boolean?

File: bayesflow/utils/keras_utils.py [hunk 3]
```diff
@@ -1,9 +1,29 @@
...
+def resolve_seed(seed):
+    """Convert an integer seed to a SeedGenerator; pass a SeedGenerator or None through unchanged."""
+    if isinstance(seed, int):
+        return keras.random.SeedGenerator(seed)
+    return seed
+
+
+def keras_multinomial(
+    num_samples: int, probs: Sequence[float], *, seed: int | keras.random.SeedGenerator | None = None
+):
```
Comment (LarsKue) [inline bayesflow/utils/keras_utils.py]: needs doc

File: bayesflow/approximators/ensemble_approximator.py [hunk 4]
```diff
@@ -268,23 +270,27 @@ def sample(
...
 
+        seed_generator = resolve_seed(seed)
+
         weights = self._resolve_member_weights(member_weights)
         names = tuple(weights.keys())
-        probs = np.fromiter(weights.values(), dtype=float, count=len(weights))
+        probs = np.array(list(weights.values()))
 
-        counts = np.random.multinomial(num_samples, probs)
+        counts = keras_multinomial(num_samples, probs, seed=seed_generator)
```
Comment (LarsKue) [inline bayesflow/approximators/ensemble_approximator.py]: This is called only in two places, and I don't see the need for the utility function here yet. I think the code would be more explicit, if we remove the utility function and repeat ourselves a little bit instead (notably not following DRY).

File: bayesflow/distributions/diagonal_student_t.py [hunk 5]
```diff
@@ -111,18 +112,19 @@ def log_prob(self, samples: Tensor, *, normalize: bool = True) -> Tensor:
         return result
 
     @allow_batch_size
-    def sample(self, batch_shape: Shape) -> Tensor:
+    def sample(self, batch_shape: Shape, seed: int | keras.random.SeedGenerator | None = None) -> Tensor:
```
Comment (LarsKue) [inline bayesflow/distributions/diagonal_student_t.py:115]: Needs doc, particularly for the seed override. If doc is inherited from the parent class (which I am not sure of right now), then we can get away with a reference to such here, but there should be some reference to the actual doc in the code.

File: bayesflow/networks/inference/scoring/scoring_rules/mixture_score.py [hunk 6]
```diff
@@ -201,32 +206,30 @@ def sample(self, batch_shape: Shape, **estimates: Tensor) -> Tensor:
         Tensor
             Samples with shape (batch_size, num_samples, ...).
         """
-        if len(batch_shape) != 2:
-            raise ValueError("`batch_shape` must be a tuple (batch_size, num_samples).")
+        assert len(batch_shape) == 1  # approximator.helpers.Sampler makes sure that batch_shape is flat
```
Comment (LarsKue) [inline bayesflow/networks/inference/scoring/scoring_rules/mixture_score.py]: Prefer raising exceptions with error messages over plain `assert`

File: bayesflow/workflows/basic_workflow.py [hunk 7]
```diff
@@ -213,446 +213,141 @@ def default_adapter(
...
-            Additional keyword arguments passed to the underlying `_fit` method.
+            If set to `"infer"` (default), the structural dimensions are inferred from the `inference_conditions`.
+            In that case, all non-vector dimensions except the last (channel) dimension are treated as structural
+            dimensions. For example, if the final `inference_conditions` have shape `(batch_size, time, channels)`,
+            then `sample_shape` is inferred as `(time,)`, and the generated samples will have shape
+            `(num_conditions, num_samples, time, target_dim)`.
+        seed : int, keras.random.SeedGenerator, or None, optional
+            Seed for reproducible sampling. An integer is converted to a ``keras.random.SeedGenerator``
+            and shared across all stochastic operations in the call. A ``SeedGenerator`` is passed through
+            as-is. If ``None`` (default), each component uses its own instance seed generator or a global one.
```
Comment (LarsKue) [inline bayesflow/workflows/basic_workflow.py]: Unspecific, see my other comment

File: tests/test_approximators/test_ensemble_approximator/test_sample.py [hunk 8]
```diff
@@ -20,3 +20,16 @@ def test_approximator_sample(ensemble_approximator, simulator, batch_size, adapt
 
     for samples_value in samples.values():
         assert isinstance(samples_value, np.ndarray)
+
+    samples_seed42_1 = ensemble_approximator.sample(num_samples=2, conditions=data, seed=42)
+    samples_seed42_2 = ensemble_approximator.sample(num_samples=2, conditions=data, seed=42)
+
+    for key in samples.keys():
+        assert np.allclose(
```
Comment (LarsKue) [inline tests/test_approximators/test_ensemble_approximator/test_sample.py]: use `np.testing.assert_allclose` for better diagnostics

File: tests/test_approximators/test_ensemble_approximator/test_sample.py [hunk 9]
```diff
@@ -20,3 +20,16 @@ def test_approximator_sample(ensemble_approximator, simulator, batch_size, adapt
...
+            samples_seed42_1[key],
+            samples_seed42_2[key],
+        ), "samples differ for identical seed"
+        assert not np.allclose(
```
Comment (LarsKue) [inline tests/test_approximators/test_ensemble_approximator/test_sample.py]: If we also want to use `np.testing.assert_allclose` here, we could do something like:

```py3
with pytest.raises(AssertionError):
    np.testing.assert_allclose(x, y)
```

File: tests/test_approximators/test_scoring_rule_approximator/test_sample.py [hunk 10]
```diff
@@ -47,3 +47,16 @@ def test_sample(scoring_rule_approximator_any, simulator, batch_size, num_sample
...
+
+    for key in samples_merged.keys():
+        assert np.allclose(
+            samples_seed42_1[key],
+            samples_seed42_2[key],
+        ), f"{key}: samples differ for identical seed"
+        assert not np.allclose(
+            samples_seed42_1[key],
+            samples_merged[key],
+        ), f"{key}: samples do not differ in unseeded case"
```
Comment (LarsKue) [inline tests/test_approximators/test_scoring_rule_approximator/test_sample.py]: see my other comment

File: tests/test_networks/test_sample.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    def leaves(x):
+        return keras.tree.flatten(x)
+
+    for s1, s2, s_unseeded in zip(leaves(samples_seed42_1), leaves(samples_seed42_2), leaves(samples_no_seed)):
+        arr1 = keras.ops.convert_to_numpy(s1)
+        arr2 = keras.ops.convert_to_numpy(s2)
+        arr_unseeded = keras.ops.convert_to_numpy(s_unseeded)
+
+        assert np.allclose(arr1, arr2), f"samples differ for identical seed ({inference_network})"
+        assert not np.allclose(arr1, arr_unseeded), f"seeded and unseeded samples are identical ({inference_network})"
```
Comment (LarsKue) [inline tests/test_networks/test_sample.py]: see my other comment

**Review by LarsKue — CHANGES_REQUESTED** [review]
Thank you for these comprehensive and progressive changes! I noted a few things I would like to see improved, but overall this is looking great already.

File: bayesflow/distributions/diagonal_student_t.py (hunk 5 above)
Comment (han-ol) [reply bayesflow/distributions/diagonal_student_t.py:115]: Can confirm that docstrings are inherited, both for help text and on sphinx docs (e.g. here: https://bayesflow.org/v2.0.10/api/bayesflow.distributions.DiagonalNormal.html#bayesflow.distributions.DiagonalNormal.get_config).

In this case do you think a reference (e.g. by comment) is necessary? If so, how would it look like?

File: bayesflow/networks/inference/diffusion/diffusion_model.py [hunk 12]
```diff
@@ -700,7 +702,7 @@ def score_fn(time, xz):
                 score_fn=score_fn,
                 noise_schedule=self.noise_schedule,
                 state=state,
-                seed=self.seed_generator,
+                seed=seed or self.seed_generator,
```
Comment (han-ol) [inline bayesflow/networks/inference/diffusion/diffusion_model.py]: will move the or case to definition of seed (so pattern is exactly as in other networks)

File: bayesflow/approximators/scoring_rule_approximator.py (hunk 2 above)
Comment (han-ol) [reply bayesflow/approximators/scoring_rule_approximator.py]: no reason. Will revert removing callable.

File: bayesflow/approximators/continuous_approximator.py (hunk 1 above)
Comment (han-ol) [reply bayesflow/approximators/continuous_approximator.py]: This just depends on what the components end up doing. Would you be in favor of removing the "or a global one"? Then we are making a normative statement to always use instance seed generators. Currently, it is descriptive, as in describing what can happen here.

I am going with the normative proposal (removing "or a global one") now, let me know if you would prefer it differently. Else, feel free to resolve.

File: bayesflow/workflows/basic_workflow.py (hunk 7 above)
Comment (han-ol) [reply bayesflow/workflows/basic_workflow.py]: see answer in other comment

**Comment by han-ol** [comment]
Thank you for your review! If you agree with how I addressed the comments and the tests are still passing this is ready for merging.

File: bayesflow/distributions/diagonal_student_t.py (hunk 5 above)
Comment (LarsKue) [reply bayesflow/distributions/diagonal_student_t.py:115]: Something like `See :py:class:...`, super simple

**Review by LarsKue — APPROVED** [review]
LGTM!
